
### Marking Rubric (which may award partial credit) 

- [0.1 points]: All relevant ChatBot summaries [including link(s) to chat log histories if you're using ChatGPT] are reported within the notebook
- [0.3 points]: Well-communicated, clear demonstration of the "model building" process and techniques of "Question 4"
- [0.3 points]: Well-communicated, clear demonstration of the "model building" process and techniques of "Question 7"
- [0.3 points]: Well-communicated, clear demonstration of the "model building" process and techniques of "Question 9"


### 1. Explain succinctly in your own words (but working with a ChatBot if needed)...<br>

1. the difference between **Simple Linear Regression** and **Multiple Linear Regression**; and the benefit the latter provides over the former


2. the difference between using a **continuous variable** and an **indicator variable** in **Simple Linear Regression**; and these two **linear forms**


3. the change that happens in the behavior of the model (i.e., the expected nature of the data it models) when a single **indicator variable** is introduced alongside a **continuous variable** to create a **Multiple Linear Regression**; and these two **linear forms** (i.e., the **Simple Linear Regression** versus the **Multiple Linear Regression**)


4. the effect of adding an **interaction** between a **continuous** and an **indicator variable** in **Multiple Linear Regression** models; and this **linear form**


5. the behavior of a **Multiple Linear Regression** model (i.e., the expected nature of the data it models) based only on **indicator variables** derived from a **non-binary categorical variable**; this **linear form**; and the necessarily resulting **binary variable encodings** it utilizes
       
<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> _The requested **linear forms** are **equations**, and answers must include **equations** along with the explanations that interpret the **linear forms**. Write you **linear forms** using a style something like_
> 
> - _"outcome = $\beta_0$ + $\beta_A$ predictorA + $\beta_B$ 1(predictorB)"_ 
> - _where the "1(.)" notation is for indicator variables_ 
> - _or feel free to use an similar alternative if a ChatBot provides you with another notation you think is clearer and like better if you prefer_
>
> _DO INCLUDE the **intercept** in your **linear forms**. You don't have to include notation related to the **error** term since this is essentially always assumed (and, actually, we usually don't even bother to include the **intercept** in such shorthand specifications either, for the same reason), but don't forget to include the **intercept** here this time (for practice). The modeling **assumptions** do not need to be addressed beyond this, but explanations will likely address the number of variables and the essential use-case (perhaps illustrated through examples) the different models imply._    
> 
> _Answers to the final question above should address the notion of a "baseline" group and it's role for **model interpretation**, why "number of categories minus one" **indicator variables** are used to represent the original **categorical variable**, and the relationship between the **binary** and **categorical variables** that are relevant for this model specification. An example use-case would likely be helpful for illustration here._ 
>    
> ---
> 
> _Don't forget to ask for summaries of all your different ChatBot sessions and organize and paste these into your homework notebook (including link(s) to chat log histories if you're using ChatBot); but, if you're using the STA130 custom NBLM ChatBot, you'll only be able to ask for summaries, of course!_ 
    
</details> 

1. Simple linear regression vs Multiple linear regression:
    Simple linear regression models the relationship between one dependent variable and one independent variable while multiple linear regression includes more than one independant variable
2. A continuous Variable represents a range of values, like age or height. The result is a smooth line across values. An indicator (Binary) Variable: Has two categories (e.g., male vs. female, coded as 0 and 1). 
3. Adding an indicator variable (like gender) adjusts the model’s starting point (intercept) for each group, giving different baseline predictions.
4. lets the slope of 𝑥 vary by group, meaning the continuous variable has a different effect in each category
5. For categorical predictors with more than two categories (like regions: North, South, East, West), we use multiple indicators. Each indicator represents one category, with one omitted as a reference.

### 2. Explain in your own words (but working with a ChatBot if needed) what the specific (outcome and predictor) variables are for the scenario below; whether or not any meaningful interactions might need to be taken into account when predicting the outcome; and provide the linear forms with and without the potential interactions that might need to be considered<br>

> Imagine a company that sells sports equipment. The company runs advertising campaigns on TV and online platforms. The effectiveness of the TV ad might depend on the amount spent on online advertising and vice versa, leading to an interaction effect between the two advertising mediums.    

1. Explain how to use these two formulas to make **predictions** of the **outcome**, and give a high level explaination in general terms of the difference between **predictions** from the models with and without the **interaction** 

2. Explain how to update and use the implied two formulas to make predictions of the outcome if, rather than considering two continuous predictor variables, we instead suppose the advertisement budgets are simply categorized as either "high" or "low" (binary variables)    
    
<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> _While working on this question, it's important to clearly understand the (**outcome** and **predictor**) **variables** under consideration, and they way they are being considered. Similarly to the previous (first) question of this homework assignment, this question requires the **equations** of the indicated **linear forms** and an explanation of their **interpretation** and use. What is different here is that the **interactions** being considered are between two **continuous variables** or two **binary variables** (for a total of four **equations** under consideration where two include the **interactions** and two do not)._
>
> _The way an **interaction** actually appears and works in the context of the linear form of a **multiple linear regression** model may not necessarily be immediately intuitive, as it is actually in fact somewhat subtle and tricky. Of course, an **interaction** is when the relationship of one **predictor variable** with the **outcome variable** depends on the value of another different **predictor variable**, so the impact of one **predictor variable** changes based on the presence or magnitude of another **predictor variable**. But are you sure you know what this means in the context of the **linear form** of a **multiple linear regression** model?_
>
> - _Imagine you're selling smoothies, the taste of your smoothie depends on the ingredients included in the smoothie, and there are two ingredients (bananas and strawberries) in the smoothie._
>     - _Adding more bananas into the smoothie will of course increase the "banana" flavor of the smoothie, and vice-versa for strawberries..._
>     - _But is this "banana" influence on the flavor always constant, or can it change depending on the absolute amount of strawberries in the smoothie?_ 
>     - _If the "banana" flavor influence is constant and does not depend on the  absolute amount of strawberries in the smoothie, then there is no **interaction** and the **linear form** of the model is $\beta_b b_i + \beta_s s_i$ and the model is said to be only **additive**._
>     - _But if the "banana" flavor influence does depend on the absolute amount of strawberries in the smoothie, then there IS an **interaction** and the **linear form** of the model is $\beta_b b_i + \beta_s s_i + \beta_{bs} (b_i\times s_i)$ and the model is said to be **synergistic**._
>         
> _These **linear forms** show that either bananas and strawberries do not have any **synergistic interaction** and contribute to the flavor independently; or, they do have a **synergistic interaction** and there is an interesting interplay between bananas and strawberries in the way they influence the taste of the smoothie._ 
> 
> - _So, if there is no **interaction**, then the effect of adding more bananas on the taste of the smoothie will always be the same, no matter how many strawberries you put in. So the effect of bananas on the smoothie is the same whether you add a lot of strawberries or just a few: $\beta_b b_i + \beta_s s_i$_
> - _Or, on the other hand, if there is an **interaction**, then the effect of adding bananas (on the smootie flavor) will be different depending on how many strawberries there currently are in the smoothie: $\beta_b b_i + \beta_s s_i + \beta_{bs} (b_i\times s_i)$_
> 
> _In this case, the right answer is probably that the **linear form** with the **interaction** is correct. This is because the flavor probably depends on the relative amount of bananas and strawberries in the smoothie; so, the effect of adding a fixed amount of bananas to the smoothie probalby depends on the absolute amount of strawberries that are in the smoothie._
> 
> _Again, because understanding **interactions** in the context of **linear forms** is somewhat subtle and tricky and indeed not necessarily obviously intuitive, let's think about this a bit more. And we can simplify the concept a little bit by considering how this **interaction** would actually technically work in a **linear form** if we just had **binary indicator variables**._
>         
> - _To consider the smootie example in terms of binary variables, suppose that if both fruits are added to the smootie, they will be added in the same amount. So the smoothie will be made with either just bananas, just strawberries, or both (or neither and you won't make a smoothie)._ 
>     - _The question regarding an **interaction** then is, is the influence of the ingredients on the taste of the smoothie **additive** or **synergistic**? That is, does the way bananas affects the flavor of the smoothie change depending on the inclusion or exclusion of strawberries in the smoothie?_
>     - _**Additive** $\beta_b 1_{[b_i=1]}(b_i) + \beta_s 1_{[s_i=1]}(s_i)$ means there are three different flavors but they are explained by just two **parameters**: banana $\beta_b$, strawberry $\beta_s$, and banana-strawberry $\beta_b+\beta_s$_
>     - _**Synergistic** $\beta_b 1_{[b_i=1]}(b_i) + \beta_s 1_{[s_i=1]}(s_i) + \beta_{bs}(1_{[b_i=1]}(b_i) \times 1_{[s_i=1]}(s_i))$ means there are of course again three different flavors, but this time they are explained by three **parameters**: banana $\beta_b$, strawberry $\beta_s$, and banana-strawberry $\beta_b+\beta_s + \beta_{bs}$, which indicates that the flavor is "more than just sum of its parts", meaning there is a **synergistic interaction** and there is an interesting interplay between bananas and strawberries in the way they influence the taste of the smoothie_
>     
> _As the **additive** and **synergistic** versions of the **linear form** of the two **binary indicator variables** context shows, we don't need an interaction to make different predictions for different combinations of things. Instead, what these show is that the prediction will either be **additive** and "just the sum of it's parts" or **synergistic** (**interactive**) and "more than just sum of its parts"._
>    
> ---
> 
> _Don't forget to ask for summaries of all your different ChatBot sessions and organize and paste these into your homework notebook (including link(s) to chat log histories if you're using ChatBot); but, if you're using the STA130 custom NBLM ChatBot, you'll only be able to ask for summaries, of course!_ 
      
</details>


In this scenario, sales revenue is the outcome variable used to gauge advertising effectiveness, with TV and online advertising budgets as predictor variables. These budgets can be either quantitative or categorical (e.g., “high” or “low”). An interaction between the budgets may be meaningful: for instance, high online spending could amplify the effect of TV ads, while low online spending might lessen it.

Without an interaction, the model assumes that TV and online budgets independently contribute to sales, represented by:

Sales=
𝛽
0
+
𝛽
1
(
TV budget
)
+
𝛽
2
(
Online budget
)
+
𝜖
Sales=β 
0
​
 +β 
1
​
 (TV budget)+β 
2
​
 (Online budget)+ϵ
However, including an interaction term captures combined effects:

Sales = 
𝛽
0
+
𝛽
1
(
TV budget
)
+
𝛽
2
(
Online budget
)
+
𝛽
3
(
TV budget
×
Online budget
)
+
𝜖
Sales=β 
0
​
 +β 
1
​
 (TV budget)+β 
2
​
 (Online budget)+β 
3
​
 (TV budget×Online budget)+ϵ
where 
𝛽
3
β 
3
​
  adjusts for cases where both budgets are high.

When using high/low categories, the model without interaction is:

Sales =
𝛽
0
+
𝛽
1
(
High TV
)
+
𝛽
2
(
High Online
)
+
𝜖
Sales=β 
0
​
 +β 
1
​
 (High TV)+β 
2
​
 (High Online)+ϵ
and with interaction:

Sales =
𝛽
0
+
𝛽
1
(
High TV
)
+
𝛽
2
(
High Online
)
+
𝛽
3
(
High TV
×
High Online
)
+
𝜖
Sales=β 
0
​
 +β 
1
​
 (High TV)+β 
2
​
 (High Online)+β 
3
​
 (High TV×High Online)+ϵ
The additive model is simpler, assuming effects are independent, while the interaction model provides more nuanced predictions, adjusting for combined effects.

### 4. Explain the apparent contradiction between the factual statements regarding the fit below that "the model only explains 17.6% of the variability in the data" while at the same time "many of the *coefficients* are larger than 10 while having *strong* or *very strong evidence against* the *null hypothesis* of 'no effect'"<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>
    
> _How do we simultaneously interpret **hypothesis testing** results regarding **coefficient estimates** based on **p-values** and **R-squared** "the proportion of variation in (outcome) $y$ explained by the model ($\hat y_i$)"? How can both be meaningfully understood at the same time? Do they address different aspects of a model?_
>    
> _As introduced in the previous homework, **R-squared** is_
>
> _$$R^2 = 1 - \frac{\sum_{i=1}^n(y_i-\hat y)^2}{\sum_{i=1}^n(y_i-\bar y)^2}$$_
>    
> _which describes the **explanatory power** of a model; whereas, **p-values** allow us to characterize **evidence against** a **null hypothesis**, and **coefficients** in a **multiple linear regression** context allow us to interpret the relationship between the **outcome** and a **predictor variable** "with all other **predictor variables** 'held constant'". Are these concepts thus contradictory or conflictual in some manner?_

|p-value|Evidence|
|-|-|
|$$p > 0.1$$|No evidence against the null hypothesis|
|$$0.1 \ge p > 0.05$$|Weak evidence against the null hypothesis|
|$$0.05 \ge p > 0.01$$|Moderate evidence against the null hypothesis|
|$$0.01 \ge p > 0.001$$|Strong evidence against the null hypothesis|
|$$0.001 \ge p$$|Very strong evidence against the null hypothesis|
    
> _In `formula='HP ~ Q("Sp. Def") * C(Generation)'` the `Q` stands for "quote" and is needed to access column names when they have a "space" in their name, while the `C` indicates a **categorical** use of what is actually an **integer** valued column. Despite technically being **continuous** numbers, **integer** often simply indicate categories which should not necessarily be treated as an incremental **continuous predictor variable**. Remember, a model such as $\beta_0 + \beta_1 x$ means for each unit increase in $x$ the outcome increases "on average" by $\beta_1$; so, if $x$ takes on the values `1` through `6` as the `Generation` **predictor variable** here does, then this means the average value for "Generation 1" must be $\beta_0 + \beta_1$ while for "Generation 2" it must be $\beta_0 + 2\times \beta_1$ (and so on up to "Generation 6" which must be $\beta_0 + 6\times \beta_1$). This might be a very strange restriction to place on something that is really actually a **categorical predictor variable**. You can see in the given model fit below how this six-level **categorical predictor variable** is actually appropriately treated in the specification of the **linear form** using "Generation 1" for the "baseline" and **binary indicators** to model the "contrast" ("offsets") for the other "Generations"; and, how these are in turn used in the context of the **interaction** considered by the model specification._ 
    
</details>

In [1]:
import pandas as pd

url = "https://raw.githubusercontent.com/KeithGalli/pandas/master/pokemon_data.csv"
# fail https://github.com/KeithGalli/pandas/blob/master/pokemon_data.csv
pokeaman = pd.read_csv(url) 
pokeaman

#                   Name   Type 1  Type 2  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison  45      49       49   
1      2                Ivysaur    Grass  Poison  60      62       63   
2      3               Venusaur    Grass  Poison  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison  80     100      123   
4      4             Charmander     Fire     NaN  39      52       43   
..   ...                    ...      ...     ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark  80     160       60   
799  721              Volcanion     Fire   Water  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 12 columns]

In [2]:
import statsmodels.formula.api as smf

model1_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") + C(Generation) + Q("Sp. Def"):C(Generation)', data=pokeaman)
model2_spec = smf.ols(formula='HP ~ Q("Sp. Def") * C(Generation)', data=pokeaman)

model2_fit = model2_spec.fit()
model2_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.176
Model:                            OLS   Adj. R-squared:                  0.164
Method:                 Least Squares   F-statistic:                     15.27
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           3.50e-27
Time:                        00:28:29   Log-Likelihood:                -3649.4
No. Observations:                 800   AIC:                             7323.
Df Residuals:                     788   BIC:                             7379.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
===================================================================================================
                                      coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------
Intercept                          26.8971      5.246      5.127      0.000      16.599      37.195
C(Generation)[T.2]                 20.0449      7.821      2.563      0.011       4.692      35.398
C(Generation)[T.3]                 21.3662      6.998      3.053      0.002       7.629      35.103
C(Generation)[T.4]                 31.9575      8.235      3.881      0.000      15.793      48.122
C(Generation)[T.5]                  9.4926      7.883      1.204      0.229      -5.982      24.968
C(Generation)[T.6]                 22.2693      8.709      2.557      0.011       5.173      39.366
Q("Sp. Def")                        0.5634      0.071      7.906      0.000       0.423       0.703
Q("Sp. Def"):C(Generation)[T.2]    -0.2350      0.101     -2.316      0.021      -0.434      -0.036
Q("Sp. Def"):C(Generation)[T.3]    -0.3067      0.093     -3.300      0.001      -0.489      -0.124
Q("Sp. Def"):C(Generation)[T.4]    -0.3790      0.105     -3.600      0.000      -0.586      -0.172
Q("Sp. Def"):C(Generation)[T.5]    -0.0484      0.108     -0.447      0.655      -0.261       0.164
Q("Sp. Def"):C(Generation)[T.6]    -0.3083      0.112     -2.756      0.006      -0.528      -0.089
==============================================================================
Omnibus:                      337.229   Durbin-Watson:                   1.505
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2871.522
Skew:                           1.684   Prob(JB):                         0.00
Kurtosis:                      11.649   Cond. No.                     1.40e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.4e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**The apparent contradiction here between a low R^2 (17.6%) and significant, large coefficients comes from the distinction between how well the model explains overall variability and the strength or significance of individual predictors.17.6% means the model explains only a small portion of the variability in the dependent variable. this model captures some relationship between HP and the predictors (Special Defense and Generation) but leaves a large amount of variability unexplained. This suggests that other important factors affecting HP are missing from the model.**

**Large and significant coefficients with low p values indicate that while each included predictor has a measurable effect on HP their combined effect doesn't explain most of HP's variability. A significant coefficient means that there is strong statistical evidence that the predictor has an effect on HP. However, it does not mean that this effect is strong enough to account for much of the variability in HP across the entire dataset.**

**These statements can be true simultaneously because the p valu shows us whether there is evidence of a non zero effect for each predictor while R^2 tells us about the overall explanatory power of the model showing how much of HP’s variation the combined predictors account for.**

## "Post-lecture" HW [*submission along with "Pre-lecture" HW is due prior to next TUT*]


### 5. Discuss the following (five cells of) code and results with a ChatBot and based on the understanding you arrive at in this conversation explain what the following (five cells of) are illustrating<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>
    
> _Recall from the previous week's homework that the **R-squared** "the proportion of variation in (outcome) $y$ explained by the model ($\hat y_i$)" quantity (re-introduced in the previous problem) can be calculated as `np.corrcoef(y,fitted_model.fittedvalues)[0,1]**2` (as well as several other ways in the **simple linear regression** context). The **squared correlation** between the outcome $y$ and it's **fitted values** $\hat y$ is the most generally useful formulation of **R-squared** since this can be use in the **multiple linear regression** context._
> 
> _This question therefore thus addresses the question of model **generalizability** on the basis of "in sample" and "out of sample" **model performance** (measured by **R-squared**)._
> 
> - _The **squared correlation** between the **outcomes** $y$ and their **fitted values** $\hat y$ is an "in sample" **model performance** metric since the $\hat y$ "predictions" for the $y$ **outcomes** are based on using those already **observed outcomes** to fit the model to generate the $\hat y$._  
> 
> - _If we instead calculate **squared correlation** between **outcomes** $y$ that were not used to fit the model and their corresponding $\hat y$ **predictions** (which are indeed now actually **predictions** as opposed to **fitted values**), then we are now  calculating an "out of sample" **model performance** metric._
> 
> _When an "out of sample" metric performs more poorly than a comparitive "in sample" metric, then the **predictions** of the **fitted model** are not **generalizing** to data being the dataset the model is fit on. In this case we say the model is **overfit** (to the data its fit was based on). The purpose of using different **training** and **testing** datasets is to consider "in sample" versus "out of sample" **model performance** in order to try to confirm that the model is not **overfit** and that the **predictions** do indeed seem to **generalizable** beyond the dataset used for **model fitting**._
>    
> ---
> 
> _Don't forget to ask for summaries of all your different ChatBot sessions and organize and paste these into your homework notebook (including link(s) to chat log histories if you're using ChatBot); but, if you're using the STA130 custom NBLM ChatBot, you'll only be able to ask for summaries, of course!_
    
</details>

In [3]:
import numpy as np
from sklearn.model_selection import train_test_split

fifty_fifty_split_size = int(pokeaman.shape[0]*0.5)

# Replace "NaN" (in the "Type 2" column with "None")
pokeaman.fillna('None', inplace=True)

np.random.seed(130)
pokeaman_train,pokeaman_test = \
  train_test_split(pokeaman, train_size=fifty_fifty_split_size)
pokeaman_train


#                     Name   Type 1   Type 2  HP  Attack  Defense  \
370  338                  Solrock     Rock  Psychic  70      95       85   
6      6                Charizard     Fire   Flying  78      84       78   
242  224                Octillery    Water     None  75     105       75   
661  600                    Klang    Steel     None  60      80       95   
288  265                  Wurmple      Bug     None  45      45       35   
..   ...                      ...      ...      ...  ..     ...      ...   
522  471                  Glaceon      Ice     None  65      60      110   
243  225                 Delibird      Ice   Flying  45      55       45   
797  720      HoopaHoopa Confined  Psychic    Ghost  80     110       60   
117  109                  Koffing   Poison     None  40      65       95   
409  373  SalamenceMega Salamence   Dragon   Flying  95     145      130   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
370       55       65     70           3      False  
6        109       85    100           1      False  
242      105       75     45           2      False  
661       70       85     50           5      False  
288       20       30     20           3      False  
..       ...      ...    ...         ...        ...  
522      130       95     65           4      False  
243       65       45     75           2      False  
797      150      130     70           6       True  
117       60       45     35           1      False  
409      120       90    120           3      False  

[400 rows x 12 columns]

 **Replaces missing values in the "Type 2" column with the string "None," ensuring no missing values affect the model fitting.The dataset is split into 50% training and 50% testing sets, ensuring models are evaluated on unseen data for generalization.Fixes the randomness for reproducibility.**

In [8]:
model_spec3 = smf.ols(formula='HP ~ Attack + Defense', 
                      data=pokeaman_train)
model3_fit = model_spec3.fit()
model3_fit.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     HP   R-squared:                       0.148
Model:                            OLS   Adj. R-squared:                  0.143
Method:                 Least Squares   F-statistic:                     34.40
Date:                Fri, 15 Nov 2024   Prob (F-statistic):           1.66e-14
Time:                        00:55:44   Log-Likelihood:                -1832.6
No. Observations:                 400   AIC:                             3671.
Df Residuals:                     397   BIC:                             3683.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     42.5882      3.580     11.897      0.000      35.551      49.626
Attack         0.2472      0.041      6.051      0.000       0.167       0.327
Defense        0.1001      0.045      2.201      0.028       0.011       0.190
==============================================================================
Omnibus:                      284.299   Durbin-Watson:                   2.006
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5870.841
Skew:                           2.720   Prob(JB):                         0.00
Kurtosis:                      20.963   Cond. No.                         343.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Predicts HP using Attack and Defense as predictors. Calculates R^2, which measures how well the model fits the training data. Uses the test set predictions to calculate an R^2 -like metric (
corr(𝑦,𝑦^)2corr(y, y^​ ) 2 ) that measures generalization to unseen data.**

### 6. Work with a ChatBot to understand how the *model4_linear_form* (*linear form* specification of  *model4*) creates new *predictor variables* as the columns of the so-called "design matrix" *model4_spec.exog* (*model4_spec.exog.shape*) used to predict the *outcome variable*  *model4_spec.endog* and why the so-called *multicollinearity* in this "design matrix" (observed in *np.corrcoef(model4_spec.exog)*) contribues to the lack of "out of sample" *generalization* of *predictions* from *model4_fit*; then, explain this consisely in your own works<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> _The **overfitting** observed in the previous problem is a question of **model complexity** relative to the amount of information or evidence available in a given dataset (or we could just say "the amount of data in a dataset"). The **model fit** for  `model4` resulted in an **overfit model** because the specification of its **linear form** was overly complex (relative to the the amount of available data). Indeed, `model4` is rediculously complex as can be seen from `model4_fit.summary()`. This in turn allowed the **model fit** to "detect" idiosyncratic associations spuriously present specifically in the **training** dataset but which did not **generalize** to the **testing** dataset. If a model is too **complex** then it will find and model "patterns" in a **training** dataset which are actually just accidental "noise" from the random sampling nature of the dataset. The simpler (or more parsimoneous) `model3` on the other hand was able to identify **predictive associations** in the **training** dataset which did **generalize** to the **testing** dataset. This is because `model3` only sought to understand simpler **predictive association** for which there was enough data in the **training** dataset to reliably detect and obviously identify. And these patterns were indeed sufficiently "real" in the sense that they were present and **generalized** into the **testing** dataset as well as the **training** dataset. So they could be "found" in the **training** and then used in (**generalized** to) the **testing** dataset._
> 
> _This question therefore addresses the topic of the **evidence** a given dataset provides for the **predictive associations** detected by a **fitted model**. It should be increasingly clear at this point that evidence for a model can be addressed using **coefficent hypothesis testing** in the context of **multiple linear regression**, but that examinations of "in sample" versus "out of sample" **model performance** metrics are what in fact are directly designed to address this question of **generalizability**. That said, this question introduces another consideration of **multicollinearity** as something that affects the **generalizability** of **model fits** in the **multiple linear regression** context. A good question that a ChatBot could help you understand is (a) "why is **generalizability** more uncertain if two **predictor variables** are highly **correlated**?" and (b) "why is **generalizability** more uncertain if multiple **predictor variables** are highly **multicollinear**?"_
>
> _The four code cells below are not necessary for answering this question; however, they introduce two very practical helpful tools for the **multiple linear regression** context that are immediately relevant for this question. The first is the so-called **condition number** (of a "design matrix") which provides a very simple diagnostic which can serve as a measure the degree of **multicollinearity** that is present in a **model fit**. If this number is "very large" then there is a large degree of **multicollinearity** and suggests room for doubt regarding the **generalizability** of the **fitted model**. The second tool(s) are the `center` and `scale` functions. It is best practice to "center and scale" **continuous predictor variables** (but not **indicator variables**) in the **multiple linear regression** context as is done below. While "centering and scaling" does make interpreting the predictions on the original scale of the data slighly more complicated, it also must be done in order to get a "true" evaluation of the degree of **multicollinearity** present in a **model fit** using the **condition number** of the model ("design matrix"). The examples below show that the **condition number** reported by a **fitted model** are "artificially inflacted" if "centering and scaling" is not used. Specically, they show that the **condition number** of `model3_fit` is really `1.66` (as opposed to the "very large" `343` which is reported without "centering and scaling"); whereas, the **condition number** for `model4_fit` is "very (VERY) large" irrespective of "centering and scaling", showing that the overwheling presense of **multicollinearity** in `model4_fit` is in fact a very real thing.  Indeed, we have already seen that `model4_fit` is grossly **overfit** and does not remotely **generalize** beyond its **training** dataset. Without knowing this, however, the comically large **condition number** for `model4_fit` (after "centering and scaling") makes it abundantly clear that we should have great doubts about the likely **generalizability** of `model4_fit` (even without examining specific aspects of **multicollinearity** directly or examining "in sample" versus "out of sample" **model performance** comparisions)._
>
> - _The "specific aspects of **multicollinearity**" reference above refer to understanding and attributing the detrmimental affects of specific **predictor variables** towards **multicollinearity**. This can be done using so-called **variance inflation factors**, but this is beyond the scope of STA130. We should be aware that the presence of excessive **multicollinearity** as indicated by a large **condition number** for a ("centered and scaled") **fitted model** raises grave concerns regarding the potential **generalizability** of the model._
>
> _The `np.corrcoef(model4_spec.exog)` examination of the **correlations** of a "design matrix" considered in ths problems prompt is analogous to the examination of the **correlations** present in a dataset that might considered when initially examining the **predictor variables** of a dataset, such as `pokeaman.iloc[:,4:12].corr()`. Indeed, such an examination is often the first step in examining the potential presence of **multicollinearity** among the **predictor variables** of a dataset. However, these are consideration of **pairwise correlation**, whereas **multicollinearity** generalizes this notion to the full collection of **predictor variables** together. A **condition number** for a "centered and scale" version of a **fit model** can therefore be viewed as serving the analogous purposes of a multivariate generalization of **pairwise correlation**._
>    
> ---
> 
> _Don't forget to ask for summaries of all your different ChatBot sessions and organize and paste these into your homework notebook (including link(s) to chat log histories if you're using ChatBot); but, if you're using the STA130 custom NBLM ChatBot, you'll only be able to ask for summaries, of course!_ 
    
</details>


In [ ]:
# "Cond. No." WAS 343.0 WITHOUT to centering and scaling
model3_fit.summary() 

In [ ]:
from patsy import center, scale

model3_linear_form_center_scale = \
  'HP ~ scale(center(Attack)) + scale(center(Defense))' 
model_spec3_center_scale = smf.ols(formula=model3_linear_form_center_scale,
                                   data=pokeaman_train)
model3_center_scale_fit = model_spec3_center_scale.fit()
model3_center_scale_fit.summary()
# "Cond. No." is NOW 1.66 due to centering and scaling

In [ ]:
model4_linear_form_CS = 'HP ~ scale(center(Attack)) * scale(center(Defense))'
model4_linear_form_CS += ' * scale(center(Speed)) * Legendary' 
model4_linear_form_CS += ' * scale(center(Q("Sp. Def"))) * scale(center(Q("Sp. Atk")))'
# Legendary is an indicator, so we don't center and scale that

model4_CS_spec = smf.ols(formula=model4_linear_form_CS, data=pokeaman_train)
model4_CS_fit = model4_CS_spec.fit()
model4_CS_fit.summary().tables[-1]  # Cond. No. is 2,250,000,000,000,000

# The condition number is still bad even after centering and scaling

In [ ]:
# Just as the condition number was very bad to start with
model4_fit.summary().tables[-1]  # Cond. No. is 12,000,000,000,000,000


High multicollinearity in model4’s design matrix leads to a poorly conditioned model that cannot generalize well. Although the model fits the training data, the collinearity between predictors causes unstable coefficient estimates that impair the model’s predictive accuracy on new data. Centering and scaling reduce, but do not eliminate this issue.

The model formula for model4 (model4_linear_form_CS) includes several predictor variables (e.g., Attack, Defense, Speed, Sp. Def, and Sp. Atk), interactions between them, and the Legendary indicator variable.This specification generates many new predictor columns in the design matrix (model4_spec.exog) representing combinations of centered and scaled interactions between these variables.

- Even after centering and scaling numeric predictors (which can sometimes reduce multicollinearity), the design matrix for model4 remains poorly conditioned (as indicated by a high condition number, Cond. No.).
High multicollinearity in the design matrix means that certain predictor columns are nearly linearly dependent on each other, creating redundancy.
- This redundancy complicates the model’s ability to learn reliable coefficients for each predictor, as slight changes in input data can disproportionately affect predictions.

A high condition number signals multicollinearity issues, which reduces the model's capacity to generalize. When the model is applied to new data, the prediction performance deteriorates because the model is overfitting the complex, collinear relationships within the training data

### 7. Discuss with a ChatBot the rationale and principles by which *model5_linear_form* is  extended and developed from *model3_fit* and *model4_fit*; *model6_linear_form* is  extended and developed from *model5_linear_form*; and *model7_linear_form* is  extended and developed from *model6_linear_form*; then, explain this breifly and consisely in your own words<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> _We again include the **condition number** for the "centered and scaled" version of `model7_fit` to show that **multicollinearity** does not appear to be a major concern for this model (and the same would be true regarding `model6_fit` if the analogous "centered and scaled" version of the model was considered). While it is true that the **condition number** of `15.4` observed for `model7_fit` is perhaps "large", this would not be considered "vary large"._
>
> - _Regarding **condition numbers**, a ChatBot gave me cutoffs of `<30` not a big problem, up to `<300` maybe an issue, up to `<1000` definitely **multicollinearity**, and beyond that is pretty much likely to be "serious" problems with **multicollinearity**. Personally, cutoffs around `10`, `100`, and `1000` seem about right to me._
>
> _This question addresses the **model building** exercise using both an **evidence** based approach using **coefficient hypothesis testing** as well as examinations of **generalizability** using comparisions of "in sample" versus "out of sample" **model performance** metrics. Through these tools, different models were considered, extended, and developed, finally arriving at `model7_fit`. When we feel we can improve the **model performance** in a **generalizable** manner, then all relatively underperforming models are said to be **underfit**, meaning that they do not leverage all the **predictive associations** available to improve **predictions**._
> 
> _While the previous "Question 6" above introduced and explored the impact of **multicollinearity** in the **multiple linear regression** context_ 
>     
> - _(whereby "the effects" of multiple **predictor variables** are "tangled up" and therefore do not allow the model to reliably determine contribution attributions between the **predictor variables**, which potentially leads to poor **estimation** of their "effects" in the model, which in turn is the problematic state of affairs which leads to a lack of **generalizability** in such high **multicollinearity** settings)_
> 
> _there is still the (actually even more important) consideration of the actual **evidence** of **predictive associations**. The question is whether or not there is sufficient **evidence** in the data backing up the **estimated** fit of the **linear form** specification. Quantifying the **evidence** for a **estimated** model is a separate question from the problem of **multicollinearity**, the assessment of which is actually the primary purpose of **multiple linear regression** methodology._
>    
> ---
> 
> _Don't forget to ask for summaries of all your different ChatBot sessions and organize and paste these into your homework notebook (including link(s) to chat log histories if you're using ChatBot); but, if you're using the STA130 custom NBLM ChatBot, you'll only be able to ask for summaries, of course!_ 
    
</details>    

In [ ]:
# Here's something a little more reasonable...
model5_linear_form = 'HP ~ Attack + Defense + Speed + Legendary'
model5_linear_form += ' + Q("Sp. Def") + Q("Sp. Atk")'
model5_linear_form += ' + C(Generation) + C(Q("Type 1")) + C(Q("Type 2"))'

model5_spec = smf.ols(formula=model5_linear_form, data=pokeaman_train)
model5_fit = model5_spec.fit()
model5_fit.summary()

In [ ]:
yhat_model5 = model5_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model5_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model5)[0,1]**2)

In [ ]:
# Here's something a little more reasonable...
model6_linear_form = 'HP ~ Attack + Speed + Q("Sp. Def") + Q("Sp. Atk")'
# And here we'll add the significant indicators from the previous model
# https://chatgpt.com/share/81ab88df-4f07-49f9-a44a-de0cfd89c67c
model6_linear_form += ' + I(Q("Type 1")=="Normal")'
model6_linear_form += ' + I(Q("Type 1")=="Water")'
model6_linear_form += ' + I(Generation==2)'
model6_linear_form += ' + I(Generation==5)'

model6_spec = smf.ols(formula=model6_linear_form, data=pokeaman_train)
model6_fit = model6_spec.fit()
model6_fit.summary()

In [ ]:
yhat_model6 = model6_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model6_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2)

In [ ]:
# And here's a slight change that seems to perhaps improve prediction...
model7_linear_form = 'HP ~ Attack * Speed * Q("Sp. Def") * Q("Sp. Atk")'
model7_linear_form += ' + I(Q("Type 1")=="Normal")'
model7_linear_form += ' + I(Q("Type 1")=="Water")'
model7_linear_form += ' + I(Generation==2)'
model7_linear_form += ' + I(Generation==5)'

model7_spec = smf.ols(formula=model7_linear_form, data=pokeaman_train)
model7_fit = model7_spec.fit()
model7_fit.summary()

In [ ]:
yhat_model7 = model7_fit.predict(pokeaman_test)
y = pokeaman_test.HP
print("'In sample' R-squared:    ", model7_fit.rsquared)
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2)

In [ ]:
# And here's a slight change that seems to perhas improve prediction...
model7_linear_form_CS = 'HP ~ scale(center(Attack)) * scale(center(Speed))'
model7_linear_form_CS += ' * scale(center(Q("Sp. Def"))) * scale(center(Q("Sp. Atk")))'
# We DO NOT center and scale indicator variables
model7_linear_form_CS += ' + I(Q("Type 1")=="Normal")'
model7_linear_form_CS += ' + I(Q("Type 1")=="Water")'
model7_linear_form_CS += ' + I(Generation==2)'
model7_linear_form_CS += ' + I(Generation==5)'

model7_CS_spec = smf.ols(formula=model7_linear_form_CS, data=pokeaman_train)
model7_CS_fit = model7_CS_spec.fit()
model7_CS_fit.summary().tables[-1] 
# "Cond. No." is NOW 15.4 due to centering and scaling

In [ ]:
# "Cond. No." WAS 2,340,000,000 WITHOUT to centering and scaling
model7_fit.summary().tables[-1]

Model 5 is an extension and simplification of Models 3 and 4. Model 3 used basic predictors (Attack and Defense), while Model 4 introduced many interaction terms that led to overfitting. In contrast, Model 5 simplifies this by focusing on the main effects of numerical variables such as Attack, Defense, Speed, Special Defense, and Special Attack. Additionally, it incorporates categorical variables like Pokémon Generation and Type, which help explain the variation in HP without overcomplicating the model with excessive interactions. This simplification strikes a balance between predictive power and generalizability, ensuring better performance on new data.

Model 6 refines Model 5 further by narrowing down the predictors based on statistical significance. It retains key numerical predictors (Attack, Speed, Sp. Def, Sp. Atk) and only includes significant categorical indicators, such as specific Pokémon types ("Normal" and "Water") and generations (2 and 5). This reduces noise from less informative variables and helps mitigate overfitting, ultimately improving the model’s out-of-sample performance while still maintaining good in-sample accuracy.

Model 7 introduces selective interaction terms, but in a more targeted way than Model 4, focusing on interactions between numerical predictors (Attack, Speed, Sp. Def, Sp. Atk). It retains the significant categorical predictors from Model 6 to ensure relevance without overwhelming the model with unnecessary complexity. This approach strikes a balance between model complexity and prediction accuracy, improving generalization compared to Model 4.


### 9. Work with a ChatBot to understand the meaning of the illustration below; and, explain this in your own words<br>

<details class="details-example"><summary style="color:blue"><u>Further Guidance</u></summary>

> _While we had seemed to **validate** the **generalizability** of `model7_fit` in **model building** exercise of the previous "Question 7" above, as well as the improved **model performance** of `model7_fit` comapred to `model6_fit`, the `model7_fit` model was always nonetheless more complex than `model6_fit` model (as seen by comparing their `.summary()` methods). This complexity, despite the minimal concerns regarding **multicollinearity**, should always have suggested some room for caution. This is because, as previously discussed in "Question 6" above, a complex **linear form** specification can allow a "**model fit** to 'detect' idiosyncratic associations spuriously present specifically in the **training** dataset but which did not **generalize** to the **testing** dataset." Indeed, a close look at the **p-values** in `model7_fit.summary()` will show that the **evidence** (in the data) for many of the **estimated coefficients** of `model7_fit` is in fact not very strong. In comparision, the **evidence** (in the data) for many of the **estimated coefficients** of `model6_fit.summary()` is consistently stronger._
>
> _As discussed towards the end of the commentary in the previous "Question 7" above, the primary purpose of **multiple linear regression** methodology is to allow us to assess the **evidence** (in the data) for a given **linear form** specification based on **coefficient hypothesis testing**. In this regard, then, `model6_fit` might be preferred over `model7_fit` despite the better "out of sample" **model performance** of `model7_fit` over `model6_fit`. This may not be enough to convince everyone however, so an additional consideration that might be made here is that the more simpler (more parsimoneous) nature of `model6_fit` should be preferred over `model7_fit` from the perspective of **model interpretability**. Indeed, it is quite unclear how exactly one should think about and understand a four-way **interaction** variable such as `Attack:Speed:Q("Sp. Def"):Q("Sp. Atk")` in conjunction with the whole host of the additional lower order interations. From a **model interpretability** perspective, understanding the meaning of the complex specification of `model7_fit` is "challenging" and "complicated" to say the least._
>
> - _There are also often circumstances where **model interpretability** can be MORE IMPORTANT than raw **model performance** in "out of sample" **prediction**._
> - _This is ESPECIALLY true if **predictive model performance** is relatively comparable between models of two different complexity levels. In such cases, the benefits of better **model interpretability** might provide a clear argument for the simpler (more parsimoneous) model, not to mention the additional potential benefit of more consistent improved **generalizability** over the the more complex model this might offer._
>
> _This question drives home the point that a simpler (more parsimoneous) model always offers the potential benefit of more consistent **generalizability**, not to mention **interpretability**, over more complex models. We should *ONLY* use increasingly complex models that without questions outperfrm simler models. The code below illustrates this by further additionally raising the consideration that the random **train-test** approach used above is actually not the most natural one available for our dataset, which has different "Generations". In fact, if we were actually using this model to make **predictions**, we would increasingly acquire more data over time which we would use to make **precictions** about future data which we haven't yet seen, which is what the code demonstrates. And low and behold, this exposes **generalizability** concerns that we missed when we used the dataset in an idealized way and not actually how we would use such a dataset in practice in the real world (where data would arrive sequentially, and current data is used to predict future data). These **generalizability** concerns do affect both models, but the appear to be more problematic for `model7_fit` than `model6_fit`, which is certainly a result of the increased complexity of `model7_fit` which always opens up the possibility of model **overfitting**._

<details>    

In [ ]:
model7_gen1_predict_future = smf.ols(formula=model7_linear_form,
                                   data=pokeaman[pokeaman.Generation==1])
model7_gen1_predict_future_fit = model7_gen1_predict_future.fit()
print("'In sample' R-squared:    ", model7_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model7_gen1_predict_future_fit.rsquared, "(gen1_predict_future)")
y = pokeaman[pokeaman.Generation!=1].HP
yhat = model7_gen1_predict_future_fit.predict(pokeaman[pokeaman.Generation!=1])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1_predict_future)")

In [ ]:
model7_gen1to5_predict_future = smf.ols(formula=model7_linear_form,
                                   data=pokeaman[pokeaman.Generation!=6])
model7_gen1to5_predict_future_fit = model7_gen1to5_predict_future.fit()
print("'In sample' R-squared:    ", model7_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model7)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model7_gen1to5_predict_future_fit.rsquared, "(gen1to5_predict_future)")
y = pokeaman[pokeaman.Generation==6].HP
yhat = model7_gen1to5_predict_future_fit.predict(pokeaman[pokeaman.Generation==6])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1to5_predict_future)")

In [ ]:
model6_gen1_predict_future = smf.ols(formula=model6_linear_form,
                                   data=pokeaman[pokeaman.Generation==1])
model6_gen1_predict_future_fit = model6_gen1_predict_future.fit()
print("'In sample' R-squared:    ", model6_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model6_gen1_predict_future_fit.rsquared, "(gen1_predict_future)")
y = pokeaman[pokeaman.Generation!=1].HP
yhat = model6_gen1_predict_future_fit.predict(pokeaman[pokeaman.Generation!=1])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1_predict_future)")

In [ ]:
model6_gen1to5_predict_future = smf.ols(formula=model6_linear_form,
                                   data=pokeaman[pokeaman.Generation!=6])
model6_gen1to5_predict_future_fit = model6_gen1to5_predict_future.fit()
print("'In sample' R-squared:    ", model6_fit.rsquared, "(original)")
y = pokeaman_test.HP
print("'Out of sample' R-squared:", np.corrcoef(y,yhat_model6)[0,1]**2, "(original)")
print("'In sample' R-squared:    ", model6_gen1to5_predict_future_fit.rsquared, "(gen1to5_predict_future)")
y = pokeaman[pokeaman.Generation==6].HP
yhat = model6_gen1to5_predict_future_fit.predict(pokeaman[pokeaman.Generation==6])
print("'Out of sample' R-squared:", np.corrcoef(y,yhat)[0,1]**2, "(gen1to5_predict_future)")

The code evaluates two models, model7 and model6, by predicting the "HP" (hit points) of Pokémon based on their attributes and generation. It trains the models on data from specific generations and tests them on other generations to assess their generalization ability. The in-sample R-squared measures how well the model fits the training data, while the out-of-sample R-squared shows how well it predicts Pokémon from generations not included in the training data.
- For example, model7_gen1_predict_future is trained on Generation 1 data and tested on other generations, while model7_gen1to5_predict_future is trained on data from all generations except Generation 6 and tested on Generation 6. These steps are repeated for model6. By comparing in-sample and out-of-sample R-squared values, the evaluation helps determine how well the models generalize.
- If the in-sample R-squared is much higher than the out-of-sample R-squared, it suggests overfitting, meaning the model performs well on training data but struggles with new data. If both values are high and similar, the model is well-regularized and generalizes well. If the out-of-sample R-squared is higher while the in-sample R-squared is low, it suggests underfitting, where the model doesn’t capture enough complexity in the training data.

The goal of this code is to evaluate how well the models generalize to unseen data, especially when data is divided by generations. This helps us understand how much the training data (specific to certain generations) impacts the model's ability to make accurate predictions for other generations. The results can guide decisions about model tuning and whether further steps are necessary to improve generalization.